File: NSM/auth-service/internal/entity/rbac.go
package entity

import "time"

// Role is a named, reusable bundle of permissions. OrganizationID is nil for
// built-in system roles (shared across every tenant); non-nil for a
// tenant-defined custom role.
type Role struct {
	ID             string
	OrganizationID *string
	Name           string
	Description    *string
	IsSystemRole   bool
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

// Permission is an atomic, system-defined capability: a Resource ("secret",
// "user") paired with an Action ("read", "write"). Permissions are global,
// not tenant-scoped, and — deliberately — have no repository Create/Delete
// methods; see internal/repository/rbac_repository.go.
type Permission struct {
	ID          string
	Resource    string
	Action      string
	Description *string
}

// Name reproduces the DB's generated `resource || ':' || action` column so
// callers never have to compute it twice.
func (p *Permission) Name() string {
	return p.Resource + ":" + p.Action
}

// RolePermission is the role<->permission junction row (role_permissions).
type RolePermission struct {
	RoleID       string
	PermissionID string
	GrantedAt    time.Time
}

File: NSM/frontend/src/types/rbac.ts
// Mirrors auth-service/internal/dto/rbac.go exactly.

export interface PermissionResponse {
  id: string
  resource: string
  action: string
  name: string
  description?: string
}

export interface RoleResponse {
  id: string
  organization_id?: string
  name: string
  description?: string
  is_system_role: boolean
  created_at: string
  updated_at: string
}

/** GET /v1/roles and GET /v1/roles/{roleId}'s body. */
export interface RoleWithPermissionsResponse extends RoleResponse {
  user_count: number
  permissions: PermissionResponse[]
}

export interface RoleGrantResponse {
  role_id: string
  role_name: string
  assigned_by?: string
  assigned_at: string
  expires_at?: string
}

export interface RoleGrantCreateRequest {
  role_id: string
  expires_at?: string
}
